src/Http/Routing/src/DependencyInjection/RoutingServiceCollectionExtensions.cs

```diff
@@ -111,12 +128,11 @@ public static IServiceCollection AddRouting(this IServiceCollection services)
         this IServiceCollection services,
         Action<RouteOptions> configureOptions)
     {
-        ArgumentNullException.ThrowIfNull(services);
-        ArgumentNullException.ThrowIfNull(configureOptions);
-
-        services.Configure(configureOptions);
-        services.AddRouting();
-
-        return services;
+        return services.AddRoutingCore().Configure<RouteOptions>(options =>
+        {
+            // Replace RegexErrorStubRouteConstraint with the real regex constraint.
+            options.SetParameterPolicy<RegexInlineRouteConstraint>("regex");
+            configureOptions(options);
+        });
```

Comment:
Can we revert this part of the change? I think the original code will still work if we reorder AddRouting() and Configure(configureOptions), and it properly null checks `configureOptions`. 

The fact that the ordering of AddRouting() and Configure(configureOptions) even matters feels like a bug. User code could already be calling Configure before AddRouting rather than using the AddRouting overload with a callback. And I don't love the duplication of adding the `RegexInlineRouteConstraint` in two places.

We should also add a test that verifies we throw an ArgumentNullException here.

```suggestion
        ArgumentNullException.ThrowIfNull(services);
        ArgumentNullException.ThrowIfNull(configureOptions);

        services.AddRouting();
        services.Configure(configureOptions);

        return services;
```